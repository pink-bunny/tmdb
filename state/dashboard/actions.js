import * as types from './types';

export const fetchTrendingMovies = (page = 1) => ({
  type: types.FETCH_TRENDING_MOVIES,
  page
});

export const fetchTrendingMoviesSuccess = (ids, totalItems, currentPage) => ({
  type: types.FETCH_TRENDING_MOVIES_SUCCESS,
  ids,
  totalItems,
  currentPage
});

export const fetchTrendingMoviesError = (error) => ({
  type: types.FETCH_TRENDING_MOVIES_ERROR,
  error
});

export const searchMovies = ({ search }, setErrors, setSubmitting, page = 1) => ({
  type: types.SEARCH_MOVIES,
  search,
  page,
  setErrors,
  setSubmitting
});
