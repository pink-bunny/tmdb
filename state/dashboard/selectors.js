export const isDashboardMoviesLoading = (state) => state.dashboard.loading;

export const dashboardMoviesList = (state) => {
  const refs = state.dashboard.ids;
  const list = state.data.movies;

  if (refs.length === 0) {
    return null;
  }
  return refs.map((item) => list[item]);
};

export const dashboardMoviesTotalPages = (state) => (
  state.dashboard.totalItems
);

export const dashboardMoviesCurrentPage = (state) => (
  state.dashboard.currentPage
);

export const dashboardMoviesError = (state) => state.dashboard.error;

export const dashboardMoviesSearchQuery = (state) => state.dashboard.searchQuery;
