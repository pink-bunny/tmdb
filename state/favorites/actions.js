import * as types from './types';

export const fetchFavorites = (page = 1) => ({
  type: types.FETCH_FAVORITES,
  page
});
export const fetchFavoritesSuccess = (ids, totalItems, currentPage) => ({
  type: types.FETCH_FAVORITES_SUCCESS,
  ids,
  totalItems,
  currentPage
});

export const fetchFavoritesError = (error) => ({
  type: types.FETCH_FAVORITES_ERROR,
  error
});

export const toggleToFavorite = (id, favorite, needRefetchList = false) => ({
  type: types.TOGGLE_TO_FAVORITES,
  id,
  favorite,
  needRefetchList
});

export const toggleToFavoriteSuccess = (id, isMovieInFavorites) => ({
  type: types.TOGGLE_TO_FAVORITES_SUCCESS,
  id,
  isMovieInFavorites
});
