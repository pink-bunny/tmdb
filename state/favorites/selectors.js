export const isFavoritesLoading = (state) => state.favorites.loading;

export const favoritesTotalItems = (state) => state.favorites.totalItems;

export const favoritesCurrentPage = (state) => state.favorites.currentPage;

export const favoritesError = (state) => state.favorites.error;

export const favoritesList = (state) => {
  const refs = state.favorites.ids || [];
  const list = state.data.movies || {};

  if (refs.length === 0) {
    return null;
  }
  return refs.map((item) => list[item]);
};
