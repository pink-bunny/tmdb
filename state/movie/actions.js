import * as types from './types';

export const fetchMovie = (id) => ({
  type: types.FETCH_MOVIE,
  id
});

export const fetchMovieSuccess = (id) => ({
  type: types.FETCH_MOVIE_SUCCESS,
  id
});

export const fetchMovieError = (error) => ({
  type: types.FETCH_MOVIE_ERROR,
  error
});

export const clearMovieFromState = () => ({
  type: types.CLEAR_MOVIE_FROM_STATE
});
