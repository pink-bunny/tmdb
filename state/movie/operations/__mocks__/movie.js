/* eslint-disable quote-props, quotes */
export const movieInfoResponseSuccess = {
  "data": {
    "adult": false,
    "backdrop_path": "/jiqD14fg7UTZOT6qgvzTmfRYpWI.jpg",
    "belongs_to_collection": null,
    "budget": 75000000,
    "genres": [
      {
        "id": 28,
        "name": "Action"
      }
    ],
    "homepage": "http://www.birdsofpreymovie.net/",
    "id": 495764,
    "imdb_id": "tt7713068",
    "original_language": "en",
    "original_title": "Birds of Prey (and the Fantabulous Emancipation of One Harley Quinn)",
    "overview": "After her breakup with the Joker, Harley Quinn joins forces with singer Black Canary, assassin Huntress, and police detective Renee Montoya to help a young girl named Cassandra, who had a hit placed on her after she stole a rare diamond from crime lord Roman Sionis.",
    "popularity": 223.588,
    "poster_path": "/h4VB6m0RwcicVEZvzftYZyKXs6K.jpg",
    "production_companies": [
      {
        "id": 9993,
        "logo_path": "/2Tc1P3Ac8M479naPp1kYT3izLS5.png",
        "name": "DC Entertainment",
        "origin_country": "US"
      }
    ],
    "production_countries": [
      {
        "iso_3166_1": "US",
        "name": "United States of America"
      }
    ],
    "release_date": "2020-02-05",
    "revenue": 189532223,
    "runtime": 109,
    "spoken_languages": [
      {
        "iso_639_1": "en",
        "name": "English"
      }
    ],
    "status": "Released",
    "tagline": "Mind Over Mayhem",
    "title": "Birds of Prey (and the Fantabulous Emancipation of One Harley Quinn)",
    "video": false,
    "vote_average": 6.8,
    "vote_count": 1177
  }
};
export const movieInfoResponseError = {
  "response": {
    "data": {
      "status_code": 34,
      "status_message": "The resource you requested could not be found."
    }
  }
};

export const movieImagesResponseSuccess = {
  "data": {
    "id": 495764,
    "backdrops": [
      {
        "aspect_ratio": 1.777777777777778,
        "file_path": "/jiqD14fg7UTZOT6qgvzTmfRYpWI.jpg",
        "height": 2160,
        "iso_639_1": null,
        "vote_average": 5.456,
        "vote_count": 9,
        "width": 3840
      }
    ],
    "posters": [
      {
        "aspect_ratio": 0.6666666666666666,
        "file_path": "/h4VB6m0RwcicVEZvzftYZyKXs6K.jpg",
        "height": 3000,
        "iso_639_1": "en",
        "vote_average": 5.654,
        "vote_count": 31,
        "width": 2000
      }
    ]
  }
};

export const movieCreditsResponseSuccess = {
  "data": {
    "id": 495764,
    "cast": [
      {
        "cast_id": 0,
        "character": "Harleen Quinzel / Harley Quinn",
        "credit_id": "5a69364e0e0a260d6400960a",
        "gender": 1,
        "id": 234352,
        "name": "Margot Robbie",
        "order": 0,
        "profile_path": "/pbSz7d1aURy88HGzFWng5EoFU81.jpg"
      }
    ],
    "crew": [
      {
        "credit_id": "5d15bcf550f7ca6e81144a30",
        "department": "Art",
        "gender": 0,
        "id": 1783,
        "job": "Production Design",
        "name": "K.K. Barrett",
        "profile_path": "/8TsblTWtRK28VXyr1EtaIfJWYWP.jpg"
      }
    ]
  }
};

export const movieStatesResponseSuccess = {
  "data": {
    "id": 495764,
    "favorite": false,
    "rated": false,
    "watchlist": true
  }
};
