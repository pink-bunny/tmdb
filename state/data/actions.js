import * as types from './types';
import { TOGGLE_TO_WATCHLIST_SUCCESS } from '../movie/types';

export const mergeMoviesList = (movies) => ({
  type: types.MERGE_MOVIES_LIST,
  movies
});

export const mergeListsList = (lists) => ({
  type: types.MERGE_LISTS_LIST,
  lists
});

export const toggleToWatchlistSuccess = (id, isMovieInWatchlist) => ({
  type: TOGGLE_TO_WATCHLIST_SUCCESS,
  id,
  isMovieInWatchlist
});
