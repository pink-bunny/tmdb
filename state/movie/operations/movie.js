import { createLogic } from 'redux-logic';
import { normalize, schema } from 'normalizr';

import { FETCH_MOVIE } from '../types';
import { fetchMovieSuccess, fetchMovieError } from '../actions';
import { mergeMoviesList } from '../../data/actions';
import { getSessionId } from '../../session/selectors';

const movieLogic = createLogic({
  type: FETCH_MOVIE,
  warnTimeout: 0,

  async process({ getState, httpClient, action }, dispatch, done) {
    const { id } = action;
    const state = getState();
    const sessionId = getSessionId(state);

    try {
      const { data: movieInfo } = await httpClient.get(`movie/${id}`);
      const { data: movieImages } = await httpClient.get(`movie/${id}/images`);
      const { data: movieCredits } = await httpClient.get(`movie/${id}/credits`);
      const { data: movieStates } = await httpClient.get(`movie/${id}/account_states?session_id=${sessionId}`);

      const data = {
        ...movieInfo,
        ...movieImages,
        ...movieCredits,
        ...movieStates
      };

      // normalize data
      const movieInfoSchema = new schema.Entity('movie');
      const normalizedInfo = normalize(data, movieInfoSchema);
      const { movie } = normalizedInfo.entities;

      dispatch(mergeMoviesList(movie));
      dispatch(fetchMovieSuccess(id));
    } catch (error) {
      const errorMessage = error.response.data.status_message;
      dispatch(fetchMovieError(errorMessage));
    }
    done();
  }
});

export default movieLogic;
