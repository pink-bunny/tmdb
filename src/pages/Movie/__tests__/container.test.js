import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import ConnectedMovie from '../container';
import { fetchMovie } from '../../../../state/movie/actions';
import { movieError } from '../../../../state/movie/selectors';

jest.mock('../../../../state/movie/selectors', () => ({
  isMovieLoading: jest.fn(() => false),
  movieError: jest.fn(),
  movieDetails: jest.fn()
}));

jest.mock('../../../../state/movie/actions', () => ({
  fetchMovie: jest.fn()
}));

describe('Movie container', () => {
  const store = configureStore()({});
  const requiredProps = {
    store,
    match: {
      params: {
        movieId: '1'
      }
    }
  };
  store.dispatch = jest.fn();
  const wrapper = shallow(<ConnectedMovie {...requiredProps} />); /* eslint-disable-line */
  const container = wrapper.dive().dive();
  const instance = container.instance();

  it('matches snapshot with required props', () => {
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with error', () => {
    movieError.mockImplementation(() => 'Mock error');
    const wrapperError = shallow(<ConnectedMovie {...requiredProps} />); /* eslint-disable-line */
    const containerError = wrapperError.dive().dive();

    expect(containerError).toMatchSnapshot();
  });

  it('componentDidMount() method run with movie id', () => {
    const { match } = instance.props;
    const { movieId } = match.params;
    instance.componentDidMount();

    expect(fetchMovie).toHaveBeenCalledWith(movieId);
  });
});
