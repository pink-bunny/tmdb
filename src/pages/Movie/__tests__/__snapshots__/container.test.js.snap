// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie container matches snapshot with error 1`] = `
<MovieComponent
  details={null}
  error="Mock error"
  loading={false}
/>
`;

exports[`Movie container matches snapshot with required props 1`] = `
<MovieComponent
  details={null}
  error=""
  loading={false}
/>
`;
